spec/datadog/profiling/collectors/discrete_dynamic_sampler_spec.rb

```diff
@@ -0,0 +1,226 @@
+      stats = simulate_load(duration_seconds: 60, events_per_second: 8, sampling_seconds: 0.005)
+      expect(stats[:sampling_ratio]).to be_between(0.45, 0.55)
+    end
+  end
+
+  context 'given a constant load' do
+    it "the higher the target overhead, the more we'll sample" do
+      # Warm-up to overcome initial hardcoded window
+      simulate_load(duration_seconds: 5, events_per_second: 4, sampling_seconds: 0.01)
+      # Start with an initial load of 4 eps @ 0.01s sampling time should give us a sampling
+      # probability of around 50% given our 2% overhead target (see similar test case above)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 4, sampling_seconds: 0.01)
+      expect(stats[:sampling_ratio]).to be_between(0.45, 0.55)
+
+      # We'll now increase our overhead target to 4%
+      update_overhead_target(max_overhead_target * 2)
+
+      # Warm-up to overcome initial hardcoded window
+      simulate_load(duration_seconds: 5, events_per_second: 4, sampling_seconds: 0.01)
+
+      # This should allow us to sample the entire load
+      stats = simulate_load(duration_seconds: 60, events_per_second: 4, sampling_seconds: 0.01)
+      expect(stats[:sampling_ratio]).to eq(1)
+    end
+  end
+
+  it 'disables sampling for next window if sampling overhead is deemed extremely high but relaxes over time' do
+    # Max overhead of 2% over 1 seconds means a max of 0.02 seconds of sampling. If each
+    # of our samples takes 2 full seconds, there's no way for us to sample and meet the target
+    # so probability and intervals must go down to 0.
+    maybe_sample(sampling_seconds: 2) # This will trigger a readjustment because sampling_seconds > readjust_window
+    expect(sampler._native_probability).to eq(0)
+
+    # Since that initial readjustment set probability to 0, all events in the next window will be ignored but
+    # ideally we should slowly relax this over time otherwise probability = 0 would be a terminal state (if we
+    # never sample again, we'll be "stuck" with the same sampling overhead view that determined probability = 0
+    # in the first place since no new sampling data came in). Because of that, over a large enough window, we
+    # should get some "are things still as bad as before?" probing samples.
+    #
+    # Question is: how long do we have to wait for probing samples? Intuitively, we need to build enough budget over
+    # time for us to be able to take that probing hit assuming things remain the same. Each adjustment window
+    # with no sampling activity earns us 0.02 seconds of budget. Therefore we need 100 of these to go by before
+    # we see the next probing sample.
+    stats = simulate_load(duration_seconds: 99, events_per_second: 4, sampling_seconds: 2)
+    expect(stats[:num_samples]).to eq(0)
+    stats = simulate_load(duration_seconds: 1, events_per_second: 4, sampling_seconds: 2)
+    expect(stats[:num_samples]).to eq(1)
```

Comment:
This seems... a bit too much? E.g. even on an extremely busy application, should we aim to have at least 1 sample per second or something like that?

Reply:
Hmm yeah I guess I went with a rather extreme situation to force the sampler's hand (sampling a single event taking 2 seconds which is already 2 adjustment windows) and that is which is leading to the extreme probing interval. I'll write a less extreme version of this.

Reply:
In the latest version, the test thresholds are adjusted, but I actually was thinking something a bit different.

It's more, should we tweak the sampler so that there's some minimum rate it will never fall below, at least during a 60-second profile period? E.g. it seems very extreme to decide not to have a single sample during a profiled period.

I don't think this is blocking for this PR, especially since we are planning to tweak with the thresholds for the new sampler anyway :)

Reply:
It's a kind of meta question of how hard/soft our target should be: any kind of forced minimum risks making the overhead target a mere suggestion. Bugs aside, the only way for us to arrive at a situation where we never sample in a 60-second period is if we consistently get single event sampling times that are individually above our overhead target (say the default 0.01 seconds). If this happens and you enforce a minimum of 1 sample/second, then you'd be going 60x over the defined threshold.

An alternative is to simply shutdown the allocation sampler if we detect that due to adjustments we are never reaching a minimum target (i.e. do not force a minimum but disable if a minimum cannot be met since data quality will be lousy).

Reply:
I'm thinking this effect can happen due to two main causes:

1. There was a hiccup somewhere. Machine is very busy, kernel descheduled profiler while it was sampling. Machine was suspended/live moved in a VM/snapshotted. That kinda thing.

2. We truly took a huge time to sample, and every sample we take will be as costly as this.

I think the challenge is -- how do we take 1) in stride, and still be conservative enough to do the right thing in 2) ? 

Furthermore, I kinda suspect (without data) that 1 may be more common than 2, so that's why I think it's worth trying to take it into consideration, if possible.

Reply:
Yes you raise a very good point. I have some ideas on it that I'm planning on research separately.